### PR TITLE
Rewrite execution environment

### LINKED ISSUE #1954: Revisit exec env and define error from Chain to VM
Design how chain interact with VM for invalid function call to VM and wasm script

Example:
If prepare function call get_external_data what behavior should be

### PR DESCRIPTION
**Pull request by taobun**
fixed #1954

### CODE REVIEW

File: chain/x/oracle/types/exec_env.go [hunk 1]
```diff
@@ -1,38 +1,69 @@
...
+		BaseEnv: BaseEnv{
+			request: req,
+		},
+		maxRawRequests: maxRawRequests,
+	}
+}
+
+// AskExternalData implements Owasm ExecEnv interface.
+func (env *PrepareEnv) AskExternalData(eid int64, did int64, data []byte) {
 	if int64(len(data)) > MaxRawRequestDataSize {
```
Comment (sorawit) [inline chain/x/oracle/types/exec_env.go:71]: In that case, should we panic here?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: go-owasm/api/bindings.h [hunk 2]
```diff
@@ -1,52 +1,83 @@
...
+  Error_DeserializationError = 8,
+  Error_GasCounterInjectionError = 9,
+  Error_SerializationError = 10,
+  Error_FunctionNotFoundError = 11,
+  Error_GasLimitExceedError = 12,
+  Error_NoMemoryWasmError = 13,
+  Error_MinimumMemoryExceedError = 14,
+  Error_SetMaximumMemoryError = 15,
+  Error_StackHeightInstrumentationError = 16,
+  Error_SetReturnDataWrongPeriodError = 31,
```
Comment (sorawit) [inline go-owasm/api/bindings.h]: Let's start this from 128

File: go-owasm/api/env.go [hunk 3]
```diff
@@ -8,47 +8,43 @@ import (
...
+	return writeSpan(calldata, data)
 }
 
 //export cSetReturnData
-func cSetReturnData(e *C.env_t, span C.Span) {
-	(*(*envIntl)(unsafe.Pointer(e))).ext.SetReturnData(readSpan(span))
+func cSetReturnData(e *C.env_t, span C.Span) C.GoResult {
+	err := (*(*envIntl)(unsafe.Pointer(e))).ext.SetReturnData(readSpan(span))
+	if err != nil {
+		if err == ErrSetReturnDataWrongPeriod {
```
Comment (sorawit) [inline go-owasm/api/env.go]: Let's have parseError function that takes care of all these comparisons.

File: go-owasm/cbindgen.toml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+# for detailed documentation of every option here.
+
+
+
+language = "C"
+
+
+############## Options for Wrapping the Contents of the Header #################
+
+ header = "/* (c) 2020 Bandprotocol. Licensed under Apache-2.0 */"
```
Comment (sorawit) [inline go-owasm/cbindgen.toml]: Band Protocol

File: go-owasm/src/error.rs [hunk 5]
```diff
@@ -18,5 +18,51 @@ pub enum Error {
   MinimumMemoryExceedError = 14,
   SetMaximumMemoryError = 15,
   StackHeightInstrumentationError = 16,
+
+  SetReturnDataWrongPeriodError = 31,
```
Comment (sorawit) [inline go-owasm/src/error.rs]: 128

**Review by sorawit — CHANGES_REQUESTED** [review]

File: go-owasm/src/span.rs [hunk 6]
```diff
@@ -18,6 +18,10 @@ impl Span {
         }
     }
 
+    pub fn create_writable(ptr: *mut u8, cap: usize) -> Span {
```
Comment (sorawit) [inline go-owasm/src/span.rs:21]: Can we add unit tests?

**Review by Benzbeeb — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**